blockchain/validate.go

```diff
@@ -1288,3 +1301,113 @@ func (b *BlockChain) CheckConnectBlockTemplate(block *btcutil.Block) error {
 	newNode := newBlockNode(&header, tip)
 	return b.checkConnectBlock(newNode, block, view, nil)
 }
+
+// ChainParams returns the Blockchain's configured chaincfg.Params.
+//
+// NOTE: Part of the ChainCtx interface.
+func (b *BlockChain) ChainParams() *chaincfg.Params {
+	return b.chainParams
+}
+
+// BlocksPerRetarget returns the number of blocks before retargeting occurs.
+//
+// NOTE: Part of the ChainCtx interface.
+func (b *BlockChain) BlocksPerRetarget() int32 {
+	return b.blocksPerRetarget
+}
+
+// MinRetargetTimespan returns the minimum amount of time to use in the
+// difficulty calculation.
+//
+// NOTE: Part of the ChainCtx interface.
+func (b *BlockChain) MinRetargetTimespan() int64 {
+	return b.minRetargetTimespan
+}
+
+// MaxRetargetTimespan returns the maximum amount of time to use in the
+// difficulty calculation.
+//
+// NOTE: Part of the ChainCtx interface.
+func (b *BlockChain) MaxRetargetTimespan() int64 {
+	return b.maxRetargetTimespan
+}
+
+// VerifyCheckpoint checks that the height and hash match the stored
+// checkpoints.
+//
+// NOTE: Part of the ChainCtx interface.
+func (b *BlockChain) VerifyCheckpoint(height int32,
+	hash *chainhash.Hash) bool {
+
+	return b.verifyCheckpoint(height, hash)
+}
+
+// FindPreviousCheckpoint finds the checkpoint we've encountered during
+// validation.
+//
+// NOTE: Part of the ChainCtx interface.
+func (b *BlockChain) FindPreviousCheckpoint() (HeaderCtx, error) {
+	checkpoint, err := b.findPreviousCheckpoint()
+	if checkpoint == nil {
+		// This check is necessary because if we just return the nil
+		// blockNode as a HeaderCtx, a caller performing a nil-check
+		// will fail. This is a quirk of go where a nil value stored in
+		// an interface is different from the actual nil interface.
+		return nil, nil
+	}
+
+	return checkpoint, err
+}
+
+// A compile-time assertion to ensure BlockChain implements the ChainCtx
+// interface.
+var _ ChainCtx = (*BlockChain)(nil)
+
+// ChainCtx is an interface that abstracts away blockchain parameters.
+type ChainCtx interface {
```

Comment:
nit: should we move the interface definitions to the top of the file?

Reply:
moved to new file